Mathlib/RingTheory/Flat/Basic.lean

```diff
@@ -229,6 +210,116 @@ instance of_projective [h : Projective R M] : Flat R M := by
   cases h with
     | _ e he => exact of_retract R _ _ _ _ he
 
+open BigOperators in
+/--
+Define the character module of `M` to be `M →+ ℚ ⧸ ℤ`.
+The character module of `M` is an injective module if and only if
+ `L ⊗ 𝟙 M` is injective for any linear map `L` in the same universe as `M`.
+-/
+lemma injective_characterModule_iff_rTensor_preserves_injective_linearMap :
+    Module.Injective R (CharacterModule M) ↔
+    ∀ ⦃N N' : Type v⦄ [AddCommGroup N] [AddCommGroup N'] [Module R N] [Module R N']
+      (L : N →ₗ[R] N'), Function.Injective L → Function.Injective (L.rTensor M) := by
+  simp_rw [injective_iff, rTensor_injective_iff_lcomp_surjective, Surjective, DFunLike.ext_iff]; rfl
+
+-- We have established a connection between preserving injectiveness of linear map and character
+-- module being an injective module. We use Baer's criterion to investigate this connection further.
+
+variable {R M N}
+
+/-- `CharacterModule M` is Baer iff `M` is flat. -/
+theorem iff_characterModule_baer : Flat R M ↔ Module.Baer R (CharacterModule M) := by
+  simp_rw [iff_rTensor_injective', Baer, rTensor_injective_iff_lcomp_surjective,
+    Surjective, DFunLike.ext_iff, Subtype.forall]; rfl
+
+variable (R M)
+
+theorem preserves_injective_linearMap {N' : Type*} [AddCommGroup N'] [Module R N'] [h : Flat R M]
+    (L : N →ₗ[R] N') (hL : Function.Injective L) : Function.Injective (L.rTensor M) := by
+  rw [rTensor_injective_iff_lcomp_surjective]
+  exact (iff_characterModule_baer.mp h).extension_property _ hL
+
+-- /-- Do we still want this?
```

Comment:
may be not. If for some reason, one need this, it can be combined by injectiveness of character module and Baer’s criterion

Reply:
In fact, the proof could simply be:
```
lemma CharacterModule.baer_of_ideal
    (inj : ∀ (I : Ideal R), Function.Injective (TensorProduct.lift ((lsmul R M).comp I.subtype))) :
    Module.Baer R (CharacterModule M) := by
  rw [← iff_characterModule_baer]; constructor; intro _ _; apply inj
```
(note that `inj` is exactly `Flat` without the `I.FG` restriction. The next lemma should be stated using `Flat` as well, if we want it.)

Reply:
I have deleted this lemma and a bit more.
From your golf, we know that character module of M is injective iff injectiveness of linear maps is preserved and character module of M is Baer iff M is flat.
So injectiveness of linear map is preserved iff character module of M is injective iff M is flat. I golfed your golf xd

Reply:
Thanks so much for the golfing. You made the main step of the proof “basis free”!

Reply:
> I have deleted this lemma and a bit more.

You removed a bit too much 😆; I think [this](https://github.com/leanprover-community/mathlib4/pull/9120/commits/828f854980596f2596426d173558c9d186c6e3e3#diff-e83d3f692b3122641f62ff418c23c7d4cbc8fe39501aa8f0c486341bdf62abc8L237-L239) is the main goal of this PR!

> Thanks so much for the golfing. You made the main step of the proof “basis free”!

Yeah essentially all the hard work has been done and almost nothing is left to be done in this PR. The main inspiration of the golfs is your [`dual_surjective_of_injective`](https://github.com/leanprover-community/mathlib4/pull/10767/files#diff-ca58fd7b9fb9bae8db4596677ace34eb0d822bcffcac8fe0e0cf0efaaf0b8966R82-R85). I just made it an iff and everything falls into place.

Reply:
> > I have deleted this lemma and a bit more.
> 
> You removed a bit too much 😆; I think [this](https://github.com/leanprover-community/mathlib4/pull/9120/commits/828f854980596f2596426d173558c9d186c6e3e3#diff-e83d3f692b3122641f62ff418c23c7d4cbc8fe39501aa8f0c486341bdf62abc8L237-L239) is the main goal of this PR! (It allows four different universes, similar to [Module.Baer.extension_property](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Module/Injective.html#Module.Baer.extension_property).)
> 

Indeed, re-added. It’s an interesting theoretical question about what happens when $M$ lives in a universal smaller than that of $R$, but I wonder if such situation naturally occurs.